webgrid/extensions.py

```diff
@@ -47,55 +52,113 @@ def ngettext(singular, plural, num, **variables):
         return arg_name
 
-    def get_args(self, grid, request_args):
-        # Start with request args, filtered by query string prefix if there is one
+    def get_sanitized_args(self, grid, args):
         incoming_args = []
         if grid.qs_prefix:
-            for key, values in MultiDict(request_args).lists():
+            for key, values in MultiDict(args).lists():
                 # Only include args that start with the grid's prefix
                 if not key.startswith(grid.qs_prefix):
                     continue
                 key = self.sanitize_arg_name(key, grid.qs_prefix)
                 for single_val in values:
                     incoming_args.append((key, single_val))
         else:
-            incoming_args = request_args
+            incoming_args = args
+
+        return MultiDict(incoming_args)
 
+    def get_args(self, grid, previous_args):
+        request_args = self.get_args_from_request()
         if 'dgreset' in request_args:
             if 'session_key' in request_args:
                 return MultiDict(dict(dgreset=1, session_key=request_args['session_key']))
             return MultiDict(dict(dgreset=1))
 
-        return MultiDict(incoming_args)
+        request_args = self.get_sanitized_args(grid, request_args)
+        request_args.update(previous_args)
+        return request_args
+
+    @abstractmethod
+    def get_args_from_request(self):
+        pass
+
+
+class RequestArgsLoader(GridPrefixBase, ArgsLoader):
+    """ Simple args loader for web request.
+
+    Args are usually passed through directly from the request. If the grid has a query string
+    prefix, the relevant args will be namespaced - sanitize them here and return the subset
+    needed for the given grid.
+
+    In the reset case, ignore most args, and return only the reset flag and session key (if any).
+    """
+    def get_args_from_request(self):
+        return self.manager.request_url_args()
+
+
+class RequestFormLoader(GridPrefixBase, ArgsLoader):
+    """ Simple form loader for web request.
+
+    Form values are usually passed through directly from the request. If the grid has a prefix, the
+    relevant args will be namespaced - sanitize them here and return the subset needed for the given
+    grid.
+
+    In the reset case, ignore most args, and return only the reset flag and session key (if any).
+    """
+    def get_args_from_request(self):
+        return self.manager.request_form_args()
+
+
+class RequestJsonLoader(ArgsLoader):
+    """ JSON loader for web request.
+
+    See :meth:`webgrid.types.Meta` for the expected JSON structure. The parsed arguments are
+    converted to the querystring arg format and merged with any previous args.
+    """
+    def json_to_args(self, data: Dict[str, Any]):
+        meta = types.Meta.from_dict(data)
+        return MultiDict(meta.to_args())
+
+    def get_args(self, grid, previous_args):
+        data = self.manager.request_json()
```

Comment:
What sort of error handling do we have here? I.e. if the developer passes in a request with invalid JSON attributes, how are we passing that information back? Do we just ignore any attributes that aren't part of our meta spec?

I'd suggest that if we have a parse error, bail quickly and return a response without processing any grid records. That response will need to happen at the view layer that we don't have in here yet, but one of these methods in the loader should return a specific exception that will be used for that purpose.

The grid has a `user_warnings` attribute that's also worth thinking about here. That will be useful for cases when the parsing succeeds and we will process the grid, but we need to change the sequence a little. The grid's `apply_qs_args` still flashes those warnings by default. I think it's time to deprecate the kwarg there and move that to the renderer level:
- change the renderer base class `__call__` to take *args and **kwargs, and pass them through to the render method
- change HTML's render to take an `add_user_warnings` kwarg, default True, and have it do what `apply_qs_args` does for that currently.

Reply:
Ben put in some basic error handling a while back on this, so that piece is resolved.

My warnings idea will not work, because by the time the renderer is called, the template will likely have already pulled and rendered warnings. Unless, that is, the view calls `grid.html()` and passes the result to the template instead of the template calling it itself - but there's no way to enforce that usage.

I do think it would be better for the user warnings to be opt-in rather than opt-out, but I'll leave that for another issue since that would be a backwards-compatibility break.